mm2src/coins/eth.rs

```diff
@@ -4950,3 +4996,110 @@ pub async fn get_eth_address(ctx: &MmArc, ticker: &str) -> MmResult<MyWalletAddr
         wallet_address,
     })
 }
+
+#[cfg(feature = "enable-nft-integration")]
+#[derive(Clone, Debug, Deserialize, Display, PartialEq, Serialize)]
+pub enum GetValidEthWithdrawAddError {
+    #[display(fmt = "My address {} and from address {} mismatch", my_address, from)]
+    AddressMismatchError {
+        my_address: String,
+        from: String,
+    },
+    #[display(fmt = "{} coin doesn't support NFT withdrawing", coin)]
+    CoinDoesntSupportNftWithdraw {
+        coin: String,
+    },
+    InvalidAddress(String),
+}
+
+#[cfg(feature = "enable-nft-integration")]
+fn get_valid_eth_withdraw_addresses(
+    coin_enum: MmCoinEnum,
+    from: &str,
+    to: &str,
+    token_add: &str,
+) -> MmResult<(Address, Address, Address, EthCoin), GetValidEthWithdrawAddError> {
+    let eth_coin = match coin_enum {
+        MmCoinEnum::EthCoin(eth_coin) => eth_coin,
+        _ => {
+            return MmError::err(GetValidEthWithdrawAddError::CoinDoesntSupportNftWithdraw {
+                coin: coin_enum.ticker().to_owned(),
+            })
+        },
+    };
+    let from_addr = valid_addr_from_str(from).map_err(GetValidEthWithdrawAddError::InvalidAddress)?;
+    if eth_coin.my_address != from_addr {
+        return MmError::err(GetValidEthWithdrawAddError::AddressMismatchError {
+            my_address: eth_coin.my_address.to_string(),
+            from: from.to_string(),
+        });
+    }
+    let to_addr = valid_addr_from_str(to).map_err(GetValidEthWithdrawAddError::InvalidAddress)?;
+    let token_addr = addr_from_str(token_add).map_err(GetValidEthWithdrawAddError::InvalidAddress)?;
+    Ok((from_addr, to_addr, token_addr, eth_coin))
+}
+
+#[cfg(feature = "enable-nft-integration")]
+#[derive(Clone, Debug, Deserialize, Display, EnumFromStringify, PartialEq, Serialize)]
+pub enum EthNftGasDetailsErr {
+    #[display(fmt = "Invalid fee policy: {}", _0)]
+    InvalidFeePolicy(String),
+    #[from_stringify("NumConversError")]
+    #[display(fmt = "Internal error: {}", _0)]
+    Internal(String),
+    #[display(fmt = "Transport: {}", _0)]
+    Transport(String),
+}
+
+#[cfg(feature = "enable-nft-integration")]
+impl From<web3::Error> for EthNftGasDetailsErr {
+    fn from(e: web3::Error) -> Self { EthNftGasDetailsErr::from(Web3RpcError::from(e)) }
+}
+
+#[cfg(feature = "enable-nft-integration")]
+impl From<Web3RpcError> for EthNftGasDetailsErr {
+    fn from(e: Web3RpcError) -> Self {
+        match e {
+            Web3RpcError::Transport(tr) | Web3RpcError::InvalidResponse(tr) => EthNftGasDetailsErr::Transport(tr),
+            Web3RpcError::Internal(internal) => EthNftGasDetailsErr::Internal(internal),
+        }
+    }
+}
+
+#[cfg(feature = "enable-nft-integration")]
+async fn get_eth_nft_gas_details(
```

Comment:
This function has a lot of duplicate code from eth withdraw function https://github.com/KomodoPlatform/atomicDEX-API/blob/2344ebb4cf8e040c84a67670a220f14014792163/mm2src/coins/eth.rs#L752 do you think they can be combined somehow? I really think for next PRs it would be beneficial if `Erc721` and `Erc1155` are part of `EthCoinType` https://github.com/KomodoPlatform/atomicDEX-API/blob/2344ebb4cf8e040c84a67670a220f14014792163/mm2src/coins/eth.rs#L354-L360
This will make it easier to implement swaps and ordermatching for NFTs in the future if we are planning to do that since `SwapOps`/`MarketCoinOps`/etc.. will have to be implemented for NFTs.

Reply:
Hmmm very good note. 
I think, if we add nft configs in coins conf file, it will work using [`eth_coin_from_conf_and_request`](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/lp_coins.rs#L2798) in `enable` method. So we will need to create "nft" coins configs based on info from native coins configs.
will need to create special ticker names, add types in EthCoinType and add types in [CoinProtocol](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/lp_coins.rs#L2563). of course it will lead to some refactoring. 

When I had a call with Artem, we decided that its no need in NFT in coins config for now, but now obviously we shouldn't avoid it.
Just this implementation will requiter user to enable NFT coin for specific chain. sounds a bit strange :D, but that is how it will work in my mind.
I will add todo note (add nft in coins config) in issue.

Reply:
> When I had a call with Artem, we decided that its no need in NFT in coins config for now, but now obviously we shouldn't avoid it.
> Just this implementation will requiter user to enable NFT coin for specific chain. sounds a bit strange :D, but that is how it will work in my mind.

Maybe we can add a field in `enable`/`enable_eth_with_tokens` request to enable NFTs with the platform coin. We wouldn't need to add NFT coin in coins config in this case (yeah, it doesn't make sense to add it to coin config as you said).
Adding `Erc721`/`Erc1155` to `EthCoinType` is more compicated than I anticipated because of the non fungibility property of NFTs. I think we will need to think more about how to do that, and if it will create more problems or not down the line before making this change.

Reply:
> Maybe we can add a field in enable/enable_eth_with_tokens request to enable NFTs with the platform coin. We wouldn't need to add NFT coin in coins config in this case.

yeah, I think we can add new optional parameters for NFT activation in `enable/enable_eth_with_tokens` methods and just use necessary info of platform coin right from coins config our-self.

Basically for getting access to `SwapOps`, `MarketCoinOps` we need to create EthCoin related to NFT (Erc721/Erc1155 EthCoinType) and add it in coins context, right? I dont think that adding coin with Erc721/Erc1155 coin type is difficult, we can just create EthCoin with nft related ticket and EthCoinType with platform coin info. 

The problems start when we need to impl orderbooks and swaps for nft. Users dont choose which 0,001 eth they want to get from the whole currency amount on the market. Another situation for NFT, users need to see info for each NFt in orderbook. Also we will need to add new Optianl params for NFT trading (token address, token id)? Questionable. Or we have to implement new NftOps for EthCoin. It depends on how much we have to refactor in the existing swap and market implementations.

Anyway, according current year plan we focus on wallet support. I feel when we get a stable wallet implementation it will be easier to make trading decisions.
I think we can leave wallet support implementation as it is now, bcz NFT trading is another epic task. I dont think it will be difficult to refactor NFT wallet support, during NFT swaps and ordermatch implementation.

I can add a separated from todo plans note for NFT swaps and ordermatching research.

>This function has a lot of duplicate code from eth withdraw function

Reply:
> I think we can leave wallet support implementation as it is now, bcz NFT trading is another epic task. I dont think it will be difficult to refactor NFT wallet support, during NFT swaps and ordermatch implementation.

I agree :)